M2/Macaulay2/m2/Core.m2

```diff
@@ -116,6 +116,28 @@ needs = filename -> if not filesLoaded#?filename then load filename else (
      (filepath, filetime) := filesLoaded#filename;
      if filetime < fileTime filepath then load filepath)
 
+-----------------------------------------------------------------------------
+-- Setup persistent history
+-----------------------------------------------------------------------------
+historyFilename = "~/.Macaulay2-history"
+
+if not gotarg "--no-readline" then (
+    addStartFunction(() -> (
+	    if fileExists historyFilename then readHistory historyFilename));
+    -- TODO: find a better alternative to addEndFunction, because
+    -- exiting with Ctrl+D duplicates the last line of history file,
+    -- but if we use lineNumber-1, then exit and restart miss the first
+    -- entry from the current session. Moreover, maybe we shouldn't save
+    -- exit and restart in the history anyway.
+    addEndFunction(() -> (
+	    if not fileExists historyFilename
+	    then (
+		dir := directoryName historyFilename;
+		if not fileExists dir then makeDirectory dir;
+		historyFilename << close);
```

Comment:
I don't like the added complexity here. These falls were silently failing before anyway, and if a someone used one of these (who?!) and actually cared about error checking they could have checked the integer that they return for the error. I would rather revert.

Reply:
Fair enough.  I added a lot of this with my environment variable idea in mind, but we wouldn't need to check if the directory exists if it's in either `$HOME` or `applicationDirectory()` since we already know they exist.

Reply:
I've reverted the three commits I added, but there should probably still be something that creates `historyFilename` if it doesn't exist before calling `appendHistory`.  Otherwise the history isn't saved.

Reply:
Hmm, I thought this was working when I tested it.
I just added a commit that creates the file and also adds an initial like which users might run into and learn about where the file is stored:
```m2
-- This is the beginning of your Macaulay2 log stored at /home/mahrud/.Macaulay2/history.m2
```